setup.py

```diff
@@ -2,241 +2,62 @@
-    # function.
-    original_src_find = DistUtilsExtra.auto.src_find
-
-    def src_find_with_excludes(attrs):
-        src = original_src_find(attrs)
-
-        for pattern in patterns:
-            DistUtilsExtra.auto.src_markglob(src, pattern)
-
-        return src
-
-    DistUtilsExtra.auto.src_find = src_find_with_excludes
-
-    return original_src_find
-
-
-def main():
-
-    # exclude files/folder patterns from being considered by distutils-extra
-    # this returns the original DistUtilsExtra.auto.src_find function
-    # so we can patch bit back in later
-    original_find_src = exclude_files([
-        "*.sh",
-        "ul",
-        "Dockerfile.build*",
-        "PKGBUILD.template",
-        "scripts/*",
-        "docs/*",
-        "glade",
-        "test",
-        "ulauncher.desktop.dev",
-        "requirements.txt",
-    ])
-
-    DistUtilsExtra.auto.setup(
-        name='ulauncher',
-        version='%VERSION%',
-        license='GPL-3',
-        author='Aleksandr Gornostal',
-        author_email='ulauncher.app@gmail.com',
-        description='Application launcher for Linux',
-        url='https://ulauncher.io',
-        data_files=DataFileList([
-            ('share/icons/hicolor/48x48/apps', [
-                'data/media/icons/hicolor/ulauncher.svg'
-            ]),
-            ('share/icons/hicolor/48x48/apps', [
-                'data/media/icons/hicolor/ulauncher-indicator.svg'
-            ]),
-            ('share/icons/hicolor/scalable/apps', [
-                'data/media/icons/hicolor/ulauncher.svg'
-            ]),
-            ('share/icons/hicolor/scalable/apps', [
-                'data/media/icons/hicolor/ulauncher-indicator.svg'
-            ]),
-            # for fedora + GNOME
-            ('share/icons/gnome/scalable/apps', [
-                'data/media/icons/hicolor/ulauncher.svg'
-            ]),
-            ('share/icons/gnome/scalable/apps', [
-                'data/media/icons/hicolor/ulauncher-indicator.svg'
-            ]),
-            # for ubuntu
-            ('share/icons/breeze/apps/48', [
-                'data/media/icons/ubuntu-mono-light/ulauncher-indicator.svg'
-            ]),
-            ('share/icons/ubuntu-mono-dark/scalable/apps', [
-                'data/media/icons/ubuntu-mono-dark/ulauncher-indicator.svg'
-            ]),
-            ('share/icons/ubuntu-mono-light/scalable/apps', [
-                'data/media/icons/ubuntu-mono-light/ulauncher-indicator.svg'
-            ]),
-            ('share/icons/elementary/scalable/apps', [
-                'data/media/icons/elementary/ulauncher-indicator.svg'
-            ]),
-        ]),
-        cmdclass={'install': InstallAndUpdateDataDirectory}
-    )
-
-    # unpatch distutils-extra src_find
-    DistUtilsExtra.auto.src_find = original_find_src
-
-
-if __name__ == '__main__':
-    main()
+        # Build Preferences before python package build
+        subprocess.run(["./ul", "build-preferences"], check=True)
```

Comment:
> So this means we are required to have the full preferences dev environment installed locally and can't use the pre-build dev container to first build the preferences, and then to install locally?

No, ul is just a thin wrapper that sources and runs the scripts in `scripts`. Check out the [ul code](https://github.com/friday/Ulauncher-Revived/blob/dev/ul) and verify this yourself. It only takes a second. I want to move this code from ul and scripts into setup.py though, but it needs to be a separate function so you could run it without building the python code. I didn't want to make this PR even bigger, but I guess I can do that.

> Also, since the ul script and the scripts directory is not included in the tarball, setup.py can't be used to install from the tar distribution.

As I see it you don't build from the tarball. With the current tar script can use the tarball to run `./setup.py install` with the flag **--skip-build** like [here](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=ulauncher#n33) (this is from the tarball as you can see on line 17).

The tarball setup.py produces is even the result of the build + setup scripts run to a dummy location and gzipped before deleting the dummy folder, and it doesn't include setup.py and setup.cfg. Since I want to switch to using that one I might have to find some way to extend/override the default behavior so that it includes those files only when making a tarball.

Reply:
What I meant was whether this is something we should be able to do local without installing all the tooling necessary for building the preferences panel in the host machine. Specifically something like this:
```
troycurtisjr@arafel: ~/working/oss/test
$ git clone git@github.com:friday/Ulauncher-Revived.git -b setup-rewrite
Cloning into 'Ulauncher-Revived'...
remote: Enumerating objects: 7504, done.
remote: Counting objects: 100% (554/554), done.
remote: Compressing objects: 100% (262/262), done.
remote: Total 7504 (delta 292), reused 537 (delta 287), pack-reused 6950
Receiving objects: 100% (7504/7504), 2.80 MiB | 2.34 MiB/s, done.
Resolving deltas: 100% (4919/4919), done.

troycurtisjr@arafel: ~/working/oss/test
$ cd Ulauncher-Revived

troycurtisjr@arafel: ~/working/oss/test/Ulauncher-Revived (setup-rewrite)
$ python setup.py install --user
running install
running bdist_egg
running egg_info
creating ulauncher_revived.egg-info
writing ulauncher_revived.egg-info/PKG-INFO
writing dependency_links to ulauncher_revived.egg-info/dependency_links.txt
writing requirements to ulauncher_revived.egg-info/requires.txt
writing top-level names to ulauncher_revived.egg-info/top_level.txt
writing manifest file 'ulauncher_revived.egg-info/SOURCES.txt'
reading manifest file 'ulauncher_revived.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'ulauncher_revived.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
Building Preferences.
+ yarn install
scripts/build-preferences.sh: line 18: yarn: command not found
Traceback (most recent call last):
...
subprocess.CalledProcessError: Command '['./ul', 'build-preferences']' returned non-zero exit status 127.
```
Normally I'd use `./ul dev-container` and first build the preferences in the standardized container created for that purpose. So my question is whether you think that workflow should go away?

Or perhaps you mean doing these tasks in the container, running the `python setup.py build` there works smoothly.

Ok, so in the new setup.py, to install using the tarball distro I need to use the `--skip-build` option. Is that standard though? I guess it isn't too odd since it is given in the [py3_install fedora rpm macro](https://src.fedoraproject.org/rpms/python-rpm-macros/blob/rawhide/f/macros.python3#_46). Though this type of tarball would break the way the fedora rpms are generated (the once in the fedora repos, not the locally generated ones), and it does feels a bit unexpected. I think this type of tarball is expected to be a released *source* tarball. Of course Python kinda muddies the water between source and not source, but the structure of these tarballs is such that the user expects to be able to "build a release from source". To be sure, the preferences are pregenerated and included in the tarball, but that is similar to autoconf projects where the first bootstrapping is done by the project to provide the `configure` script to run in order to build the source release.

Reply:
> What I meant was whether this is something we should be able to do local without installing all the tooling necessary for building the preferences panel in the host machine.

The only dependencies you need to install on your system is Yarn and Node. This has to be documented though. Yarn depends on Node. So installing yarn gets you all you need. The packages are installed when you run `yarn install`, which the script does. This also caches them to your home so you only have to download the same package+version once. If you want you can find out where with `yarn cache dir` and delete these with `yarn cache clean`. I think this is a lot less complicated than asking people to use containerization (which might be a new concept to them) just to get their dev environment running. I don't know about you, but if people tell me to run a series of commands/scripts I don't just do it. I find out what these commands do first.

> Normally I'd use `./ul dev-container` and first build the preferences in the standardized container created for that purpose. So my question is whether you think that workflow should go away?

I think at least it shouldn't be required. You can still do it this way. Run `./ul build-preferences` from within the container, then when you build it should detect that preferences are already built when you run `./setup.py build` or `./setup.py install`

We still need Docker/Podman to build distro releases, so it makes sense to keep part of this, but just for the maintainer scripts imo.

> Ok, so in the new setup.py, to install using the tarball distro I need to use the `--skip-build` option. Is that standard though? I guess it isn't too odd since it is given in the [py3_install fedora rpm macro](https://src.fedoraproject.org/rpms/python-rpm-macros/blob/rawhide/f/macros.python3#_46). Though this type of tarball would break the way the fedora rpms are generated (the once in the fedora repos, not the locally generated ones), and it does feels a bit unexpected. I think this type of tarball is expected to be a released _source_ tarball. Of course Python kinda muddies the water between source and not source, but the structure of these tarballs is such that the user expects to be able to "build a release from source". To be sure, the preferences are pregenerated and included in the tarball, but that is similar to autoconf projects where the first bootstrapping is done by the project to provide the `configure` script to run in order to build the source release.

Yeah. You changed my mind. I agree. The standard I am used to in general with tarballs is to include the source, not the build. We'll stick to using the custom script or a variant of it. Thanks for the feedback!